### PR TITLE
Adding Support for IAM-based Authorization.

### PR DESCRIPTION
**Pull request by grundprinzip**
This patch adds support to dynamically generate passwords
necessary to access Redshift by using the Redshfit API
GetClusterCredentials().

https://docs.aws.amazon.com/redshift/latest/APIReference/API_GetClusterCredentials.html

It intercepts the extraction of the connection arguments, if the
new additional `iam` parameter is present. It expects the
hostname parameter of the URL to be the actual cluster identifier
and will use this to lookup the DNS name and port configuration.
In addition, it updates the username and password with the
updated values retrieved from the API call.

It expects the URL for the engine to be of the following structure:

    redshift://user@cluster_identifier/database?iam=true

Testing:

This patch adds two new unit tests for parsing the URL and extracting
the correct parameters from the API calls. It adds boto3 to the Tox
dependencies for mock testing.

## Todos
- [ ] MIT compatible
- [ ] Tests
- [ ] Documentation
- [ ] Updated CHANGES.rst

### CODE REVIEW

File: sqlalchemy_redshift/dialect.py [hunk 1]
```diff
@@ -641,23 +643,85 @@ def keyfunc(column):
...
+            if "password" in cparams:
+                raise ValueError(
+                    "'password' cannot be used when connecting with"
+                    " IAM federation."
+                )
+
+            return self._do_generate_iam_auth_token(redshift, cparams)
+
+        except ImportError:
+            raise ImportError(
```
Comment (graingert) [inline sqlalchemy_redshift/dialect.py]: unreachable

File: sqlalchemy_redshift/dialect.py [hunk 2]
```diff
@@ -641,23 +643,85 @@ def keyfunc(column):
...
         default_args = {
-            'sslmode': 'verify-full',
-            'sslrootcert': pkg_resources.resource_filename(
-                __name__,
-                'redshift-ca-bundle.crt'
-            ),
+            'sslmode':
+            'verify-full',
+            'sslrootcert':
+            pkg_resources.resource_filename(__name__,
```
Comment (graingert) [inline sqlalchemy_redshift/dialect.py]: why did you move all this around?

File: sqlalchemy_redshift/dialect.py [hunk 3]
```diff
@@ -641,23 +643,85 @@ def keyfunc(column):
...
+                                            'redshift-ca-bundle.crt'),
         }
-        cargs, cparams = super(RedshiftDialect, self).create_connect_args(
-            *args, **kwargs
-        )
+        cargs, cparams = super(RedshiftDialect,
```
Comment (graingert) [inline sqlalchemy_redshift/dialect.py]: very ugly whitespace only changes

**Comment by graingert** [comment]
this doesn't work because the engine will store the password forever but the password expires after 3600 seconds (1 hour)

**Comment by graingert** [comment]
can we do this with botocore instead?

**Comment by graingert** [comment]
> this doesn't work because the engine will store the password forever but the password expires after 3600 seconds (1 hour)

I think this authentication dance needs to happen every time a connection is made - maybe caching the password for 3000 seconds

**Comment by graingert** [comment]
perhaps we need to parse the args ourselves and pass along our own  https://docs.sqlalchemy.org/en/13/core/engines.html#sqlalchemy.create_engine.params.creator

**Comment by graingert** [comment]
@zzzeek can you recommend a good place to hook this?

File: sqlalchemy_redshift/dialect.py (hunk 2 above)
Comment (grundprinzip) [reply sqlalchemy_redshift/dialect.py]: What are you using as auto-formatting settings so I don't have to manually adhere to the guide?

File: sqlalchemy_redshift/dialect.py (hunk 3 above)
Comment (grundprinzip) [reply sqlalchemy_redshift/dialect.py]: yes, will revert

File: sqlalchemy_redshift/dialect.py (hunk 1 above)
Comment (grundprinzip) [reply sqlalchemy_redshift/dialect.py]: I wanted to check if it's fine to add boto3 as a dependency, IMHO it makes sense because Redshift is at the end of the day a AWS product :)

File: sqlalchemy_redshift/dialect.py (hunk 2 above)
Comment (graingert) [reply sqlalchemy_redshift/dialect.py]: no autoformat here at the moment. For multiline forms just make sure the closing bracket is on the same column as the start of the line that opens it, and every indent is a multiple of 4 and only 4 more than the previous line

```
# good
fn(
    example,
    example,
    example,
)

# good
fn(example, example, example)

# bad
fn(
    example,
    example,
    example
)

# bad
fn(example, example, example,)

# bad
fn(
    example,
    example,
    example)

# awful
fn(example,
   example,
   example)
```

**Comment by grundprinzip** [comment]
I think the best place might be to use https://docs.sqlalchemy.org/en/13/core/events.html#sqlalchemy.events.DialectEvents

**Comment by graingert** [comment]
yeah not sure if we're supposed to listen to https://docs.sqlalchemy.org/en/13/core/events.html#sqlalchemy.events.DialectEvents.do_connect or override https://docs.sqlalchemy.org/en/13/core/internals.html#sqlalchemy.engine.default.DefaultDialect.connect

**Comment by graingert** [comment]
perhaps a cached method using https://cachetools.readthedocs.io/en/latest/#cachetools.cachedmethod and https://cachetools.readthedocs.io/en/latest/#cachetools.TTLCache

```
@cachetools.cachedmethod(...)
def _get_cluster_credentials(self, cluster_identifier, db_user): ...

def connect(self, *cargs, **cparams):
    ...
    password = self._get_cluster_credentials(cluster_identifier, db_user)
    ...
    super().connect(*newcargs, **newcparams)
```

**Comment by zzzeek** [comment]
well that document is actually incorrect, the "connect" method is actually called "on_connect()", and the doc is not that clear that this is a post-connection method.  ill update that momentarily.

the real "connect" method on DefaultDialect is probably what you want to be overriding here, here's what it looks like right now (note no docstring at the moment):

    def connect(self, *cargs, **cparams):
        return self.dbapi.connect(*cargs, **cparams)

so if you need to do some token generation and caching that's probably where it can be invoked.

**Comment by graingert** [comment]
> perhaps a cached method using [cachetools.readthedocs.io/en/latest/#cachetools.cachedmethod](https://cachetools.readthedocs.io/en/latest/#cachetools.cachedmethod) and [cachetools.readthedocs.io/en/latest/#cachetools.TTLCache](https://cachetools.readthedocs.io/en/latest/#cachetools.TTLCache)
> 
> ```python
> @cachetools.cachedmethod(...)
> def _get_cluster_credentials(self, cluster_identifier, db_user): ...
> 
> def connect(self, *cargs, **cparams):
>     ...
>     password = self._get_cluster_credentials(cluster_identifier, db_user)
>     ...
>     super().connect(*newcargs, **newcparams)
> ```

ok no need to mess with a fancy cache then the password can be stored on self and protected with a lock

**Comment by bbayles** [comment]
FWIW: The AWS API limit for the GetClusterCredentials was low enough that I kept hitting it. I added retries and backoffs, but ultimately wound up using a time-based cache.

File: sqlalchemy_redshift/dialect.py [hunk 4]
```diff
@@ -641,23 +643,85 @@ def keyfunc(column):
             'redshift_interleaved_sortkey': interleaved_sortkey,
         }
 
+    def _do_generate_iam_auth_token(self, redshift, cparams):
+        # Hostname is the cluster identifier
+        cluster_id = cparams['host']
+        response = redshift.describe_clusters(ClusterIdentifier=cluster_id)
```
Comment (myoung34) [inline sqlalchemy_redshift/dialect.py:649]: Is there a way to have this be given instead of using a describeClusters call? At scale this is a great way to get rate limited.

Comment (myoung34) [reply sqlalchemy_redshift/dialect.py:649]: i would say to do something like:

```
hostname = cparams.get('hostname')
port = cparams.get('port')
if cparams.get('host'):
    response = redshift.describe_clusters(ClusterIdentifier=cluster_id)
    ...
    hostname = ...
    port = ...
```

or similar

**Review by myoung34 — CHANGES_REQUESTED** [review]

**Comment by graingert** [comment]
I'd like this to be able to use aiobotocore so we can support the AsyncEngine

**Comment by edingroot** [comment]
Hi @grundprinzip, may I know why this PR was closed? It will be very useful to support IAM credentials natively. Thanks!

**Comment by Brooke-white** [comment]
@edingroot not sure if this fits your use case, but you can achieve this using the existing [redshift-connector](https://github.com/aws/amazon-redshift-python-driver) dialect [we added a few years back](https://github.com/sqlalchemy-redshift/sqlalchemy-redshift/pull/232)
